ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala

```diff
@@ -212,15 +212,16 @@ private[kvutils] class TransactionCommitter(
         // which are not evidenced in the transaction itself and hence the contract key state is
         // not included in the inputs.
         lazy val knownKeys: Map[DamlContractKey, Value.ContractId] =
-          commitContext.inputs.collect {
-            case (key, Some(value))
-                if value.hasContractState
-                  && value.getContractState.hasContractKey
-                  && contractIsActiveAndVisibleToSubmitter(
-                    transactionEntry,
-                    value.getContractState) =>
-              value.getContractState.getContractKey -> Conversions.stateKeyToContractId(key)
-          }
+          commitContext
+            .readAllFiltered(_.getContractId.nonEmpty)
```

Comment:
Should we consider adding a `collectFiltered` (not a good name, I know) to only do a single pass? It also looks like we're recording a larger set (all non empty contractIds) than what we actually need (only those that have a contract key).

Reply:
Yes, that should work as well. When do contract IDs not have an associated contract key? I.e., in which scenarios would cause this a difference in the returned read-set?

Reply:
I think it's a good idea to have a function that reads many inputs using a partial function. Perhaps `collectInputs` would be a decent name, as it's effectively a wrapped version of `.inputs.collect { ... }` with a side effect.

Reply:
If there is a need to filter on values, another option would be to allow filtering on them as well or to provide use-case specific encapsulations of `inputs`.

Reply:
BTW, I am fairly sure most contracts don't have keys; you have to add a contract key explicitly with the DAML `key` keyword.

Reply:
@fabiotudone-da: If you filter out a value, do you still need to mark it as read?

If so, I'd prefer if that was somehow separated in the API.

Reply:
sorry, I edited the original comment because github doesn't auto update without refreshing.
Yes, if you filter on the value, you must have read the value to be able to do that.

Reply:
> @fabiotudone-da: If you filter out a value, do you still need to mark it as read?

No, AFAIU.

Reply:
filter ON != filter OUT

Reply:
@SamirTalwar Ignore my comment, @gerolf-da is right, because the filtering is logical.

Reply:
Sorry, that's what I meant. Filtering out _based on_ values.

Reply:
Replaced by `collectInputs` that takes a partial function to do the filtering/transformation.